internal/gcs/bridge.go

```diff
@@ -0,0 +1,356 @@
+	}
+}
+
+func (call *rpc) complete(err error) {
+	call.err = err
+	close(call.ch)
+}
+
+// Err returns the RPC's result. This may be a transport error or an error from
+// the message response.
+func (call *rpc) Err() error {
+	if call.err != nil {
+		return call.err
+	}
+	return call.resp.Err()
+}
+
+// Done returns whether the RPC has completed.
+func (call *rpc) Done() bool {
+	select {
+	case <-call.ch:
+		return true
+	default:
+		return false
+	}
+}
+
+// RPC issues a synchronous RPC request. Returns immediately if the context
+// becomes done and the message is not sent.
+//
+// If allowCancel is set and the context becomes done, returns an error without
+// waiting for a response. Avoid this on messages that are not idempotent or
+// otherwise safe to ignore the response of.
+func (brdg *bridge) RPC(ctx context.Context, proc rpcProc, req requestMessage, resp responseMessage, allowCancel bool) error {
+	call, err := brdg.AsyncRPC(ctx, proc, req, resp)
+	if err != nil {
+		return err
+	}
+	var ctxDone <-chan struct{}
+	if allowCancel {
+		// This message can be safely cancelled by ignoring the response.
+		ctxDone = ctx.Done()
+	}
+	t := time.NewTimer(brdg.Timeout)
+	defer t.Stop()
+	select {
+	case <-call.ch:
+		return call.Err()
+	case <-ctxDone:
+		brdg.log.WithField("reason", ctx.Err()).Warn("ignoring response to bridge message")
+		return ctx.Err()
+	case <-t.C:
+		brdg.log.Warn("terminating bridge due to message timeout")
+		brdg.Kill()
+		<-call.ch
+		return call.Err()
+	}
+}
+
+func (brdg *bridge) recvLoopRoutine() {
+	brdg.recvLoop()
+	// Close the connection so that the writer stops.
+	brdg.conn.Close()
+	// Fail any remaining RPCs.
+	brdg.mu.Lock()
+	rpcs := brdg.rpcs
+	brdg.rpcs = nil
+	brdg.mu.Unlock()
+	for _, call := range rpcs {
+		call.complete(errBridgeClosed)
+	}
+}
+
+func readMessage(r io.Reader) (int64, msgType, []byte, error) {
+	var h [16]byte
+	_, err := io.ReadFull(r, h[:])
+	if err != nil {
+		return 0, 0, nil, err
+	}
+	typ := msgType(binary.LittleEndian.Uint32(h[:]))
+	n := binary.LittleEndian.Uint32(h[4:])
+	id := int64(binary.LittleEndian.Uint64(h[8:]))
+	if n < 16 || n > 0x10000 {
+		return 0, 0, nil, fmt.Errorf("invalid message size %d", n)
+	}
+	n -= 16
+	b := make([]byte, n)
+	_, err = io.ReadFull(r, b)
+	if err != nil {
+		if err == io.EOF {
+			err = io.ErrUnexpectedEOF
+		}
+		return 0, 0, nil, err
+	}
+	return id, typ, b, nil
+}
+
+func isLocalDisconnectError(err error) bool {
+	if o, ok := err.(*net.OpError); ok {
+		if s, ok := o.Err.(*os.SyscallError); ok {
+			return s.Err == syscall.WSAECONNABORTED
+		}
+	}
+	return false
+}
+
+func (brdg *bridge) recvLoop() {
+	br := bufio.NewReader(brdg.conn)
+	for {
+		id, typ, b, err := readMessage(br)
+		if err != nil {
+			if err == io.EOF || isLocalDisconnectError(err) {
+				brdg.log.Debug("bridge disconnected")
+			} else {
+				brdg.log.WithError(err).Error("bridge read failed")
+			}
+			return
+		}
+		brdg.log.WithFields(logrus.Fields{
```

Comment:
Can we log the activity id as well?

Reply:
Oh Ugh. That's part of the JSON Message. That is very unfortunate

Reply:
I'm not convinced that I have the logging at the right layer. We might want to move it up to the RPC layer (except in unmarshal failures), in which case we can easily log the activity ID.

Reply:
That's fine we can redo the logging as we get this running